ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -211,62 +211,69 @@ module Ldap = struct
-    let get_string k =
-      match Map.find_opt k kvps with
+    let get_string key =
+      match Map.find_opt key kvps with
       | None ->
-          Error (ldap "missing key '%s'" k)
+          Error (ldap "missing key '%s'" key)
       | Some x ->
           Ok x
     in
-    let get_string_with_default ~k ~default =
-      match get_string k with Ok x -> Ok x | Error _ -> Ok default
+    let get_string_with_default ~key ~default =
+      match get_string key with Ok x -> Ok x | Error _ -> Ok default
     in
-    let get_int of_string k =
-      let* str = get_string k in
+    let get of_string key =
+      let* str = get_string key in
       try Ok (of_string str)
-      with _ -> Error (ldap "invalid value for key '%s'" k)
+      with _ -> Error (ldap "invalid value for key '%s'" key)
     in
-    let get_int_with_default of_string ~k ~default =
-      match get_int of_string k with Ok x -> Ok x | Error _ -> Ok default
+    let get_with_default of_string ~key ~default =
+      match get of_string key with Ok x -> Ok x | Error _ -> Ok default
+    in
+    let windows_nt_time_to_unix_time x =
+      Int64.sub (Int64.div x 10000000L) 11644473600L
     in
     let default = "" in
-    let* name = get_string_with_default ~k:"name" ~default in
-    let* upn = get_string_with_default ~k:"userPrincipalName" ~default in
-    let* display_name = get_string_with_default ~k:"displayName" ~default in
-    let* user_account_control = get_int Int32.of_string "userAccountControl" in
-    let* account_expires = get_int Int64.of_string "accountExpires" in
+    let* name = get_string_with_default ~key:"name" ~default in
+    let* upn = get_string_with_default ~key:"userPrincipalName" ~default in
+    let* display_name = get_string_with_default ~key:"displayName" ~default in
+    let* user_account_control = get Int32.of_string "userAccountControl" in
+    let* account_expires = get Int64.of_string "accountExpires" in
+    let* password_expires_computed =
+      get_with_default Int64.of_string
+        ~key:"msDS-UserPasswordExpiryTimeComputed" ~default:Int64.max_int
+    in
     (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-lockouttime *)
     let* lockout_time =
-      get_int_with_default Int64.of_string ~k:"lockoutTime" ~default:0L
+      get_with_default Int64.of_string ~key:"lockoutTime" ~default:0L
     in
-    let account_expired =
-      (* see https://docs.microsoft.com/en-us/windows/win32/adschema/a-accountexpires *)
-      let windows_nt_time_to_unix_time x =
-        Int64.sub (Int64.div x 10000000L) 11644473600L
-      in
-      match account_expires with
-      | x when x = 0L || x = Int64.max_int ->
+    let is_expired zero_expired = function
+      | 0L ->
+          zero_expired
+      | i when i = Int64.max_int ->
           false
-      | i ->
+      | _ as t ->
           let expire_unix_time =
-            windows_nt_time_to_unix_time i |> Int64.to_float
+            windows_nt_time_to_unix_time t |> Int64.to_float
```

Comment:
The comparison below relies on the host's clock and the DC's clock to be in sync. We use NTP on our host (though not necessarily with the DC as clock source), so that just leaves a possible DC misconfiguration to cause clocks to be out of sync (which we've seen happen, but in that case the DC's clock should be fixed, if the clock on the host is correct). Might be worth checking for host <-> DC timedrift though in some places and warn in the logs if we detect it. Is there an API to query the DC for the time through LDAP/Kerberos?

Reply:
I found a way to query DC time before, But I cannot remember where, 

Here is actually a non-critical path, where we display domain user status, not for any critical usage, like permit access to dom0.  
The different timezone may cause the display about user information not right for a while, but will not bring critical issues.